datanommer.models/datanommer/models/__init__.py

```diff
@@ -128,7 +130,16 @@ def add(envelope):
     obj.msg = message['msg']
     obj.headers = headers
 
-    session.add(obj)
+    # Construct an INSERT ... SELECT ... which will do nothing if an
+    # entry with a matching msg_id already exists. This can happen in the
+    # event of message redelivery.
+    cols = [c for c in Message.__table__.c if c.name != 'id']
+    vals = [literal(getattr(obj, c.name)) for c in cols]
+    not_exists = ~exists(select([Message.__table__.c.i]).\
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```

Comment:
You're correct, it's to avoid a database-level error that would require the transaction to be rolled-back. In this case, it's not possible for a message to exist in the database without the associated user/package. However, if there are two message consumers populating the same database, then user/package creation will race between those two consumers. The INSERT ... SELECT construct does the existence check and insert atomically, avoiding that race condition.